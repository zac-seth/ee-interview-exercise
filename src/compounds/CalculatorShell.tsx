import { FC } from "react";

import { Grid, LayoutDirection } from "elements";

type CalculatorSections = {
  display: JSX.Element;
  controls: JSX.Element;
  functions: JSX.Element;
  numbers: JSX.Element;
};

interface AssignSections {
  (): CalculatorSections;
}

export type CalculatorShellProps = {
  children: AssignSections;
};

export const CalculatorShell: FC<CalculatorShellProps> = ({ children }) => {
  const { display, controls, functions, numbers } = children();

  return (
    <Grid
      data-testid="calculator-shell-container"
      items={[1, 4, 1]}
      layoutDirection={LayoutDirection.COLUMN}
    >
      {display}
      <Grid items={[3, 1]} data-testid="calculator-shell-inner">
        {numbers}
        {functions}
      </Grid>
      {controls}
    </Grid>
  );
};
