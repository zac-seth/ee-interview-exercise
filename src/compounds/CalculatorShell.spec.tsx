import { render, screen } from "@testing-library/react";

import { CalculatorShell } from "./CalculatorShell";

const provideSections = () => ({
  controls: <p>section: controls</p>,
  display: <p>section: display</p>,
  functions: <p>section: functions</p>,
  numbers: <p>section: numbers</p>
});

describe(`CalculatorShell`, () => {
  it(`Should render the shell`, () => {
    render(
      <CalculatorShell>
        {provideSections}
      </CalculatorShell>
    );

    const grids = screen.getAllByTestId(`grid-layout`);

    expect(grids[0]).toMatchSnapshot();
  });

  it(`Should render the sections correctly`, () => {
    render(
      <CalculatorShell>
        {provideSections}
      </CalculatorShell>
    );

    const sections = screen.getAllByText(`section:`, { exact: false });

    expect(sections[0]).toHaveTextContent(`section: display`);
    expect(sections[1]).toHaveTextContent(`section: numbers`);
    expect(sections[2]).toHaveTextContent(`section: functions`);
    expect(sections[3]).toHaveTextContent(`section: controls`);
  });
});
