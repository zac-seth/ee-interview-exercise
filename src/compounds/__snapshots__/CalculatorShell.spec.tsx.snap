// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalculatorShell Should render the shell 1`] = `
.c0 {
  display: grid;
  grid-gap: 0.25rem;
  margin: 0;
  padding: 0;
  grid-template-rows: 1fr 4fr 1fr;
}

.c1 {
  display: grid;
  grid-gap: 0.25rem;
  margin: 0;
  padding: 0;
  grid-template-columns: 3fr 1fr;
}

<div
  class="c0"
  data-testid="grid-layout"
>
  <p>
    section: display
  </p>
  <div
    class="c1"
    data-testid="grid-layout"
  >
    <p>
      section: numbers
    </p>
    <p>
      section: functions
    </p>
  </div>
  <p>
    section: controls
  </p>
</div>
`;
